Snowflake.Data/Core/Converter/JsonToStructuredTypeConverter.cs

```diff
@@ -0,0 +1,423 @@
+            {
+                var value = json.Value<string>();
+                return s_timeConverter.Convert(value, SFTimestampType.TIME, fieldType);
+            }
+            if (IsDate(fieldMetadata))
+            {
+                var value = json.Value<string>();
+                return s_timeConverter.Convert(value, SFTimestampType.DATE, fieldType);
+            }
+            if (IsBinaryMetadata(fieldMetadata))
+            {
+                var value = json.Value<string>();
+                if (fieldType == typeof(byte[]))
+                {
+                    return SFDataConverter.HexToBytes(value);
+                }
+                throw new Exception($"Could not read BINARY into {fieldType}");
+            }
+            if (IsObjectMetadata(fieldMetadata)) // semi structured object
+            {
+                return json.ToString();
+            }
+            if (IsArrayMetadata(fieldMetadata)) // semi structured array
+            {
+                return json.ToString();
+            }
+            if (IsVariantMetadata(fieldMetadata))
+            {
+                return json.ToString();
+            }
+            throw new Exception($"Could not read {fieldMetadata.type} into {fieldType}");
+        }
+
+        private static object ConvertToArray(Type type, Type elementType, List<FieldMetadata> fields, JToken json)
+        {
+            if (json == null || json.Type == JTokenType.Null || json.Type == JTokenType.Undefined)
+            {
+                return null;
+            }
+            if (json.Type != JTokenType.Array)
+            {
+                throw new Exception("Json is not an array");
+            }
+            var jsonArray = (JArray)json;
+            var arrayType = MakeArrayType(type, elementType);
+            var result = (IList) Activator.CreateInstance(arrayType, jsonArray.Count);
+            var elementMetadata = fields[0];
+            for (var i = 0; i < jsonArray.Count; i++)
+            {
+                result[i] = ConvertToStructuredOrUnstructuredValue(elementType, elementMetadata, jsonArray[i]);
+            }
+            if (type != arrayType)
+            {
+                var listType = type.IsAbstract ? typeof(List<>).MakeGenericType(elementType) : type;
+                var list = (IList) Activator.CreateInstance(listType);
+                for (int i = 0; i < result.Count; i++)
+                {
+                    list.Add(result[i]);
+                }
+                return list;
+            }
+
+            return result;
+        }
+
+        private static object ConvertToMap(Type type, Type keyType, Type valueType, List<FieldMetadata> fields, JToken json)
+        {
+            if (keyType != typeof(string)
+                && keyType != typeof(int) && keyType != typeof(int?)
+                && keyType != typeof(long) && keyType != typeof(long?))
+            {
+                throw new Exception($"Unsupported key type of dictionary {keyType}");
```

Comment:
what dictionary? Did you mean: "Unsupported type of key for the conversion destination's map?

Reply:
I changed to "Unsupported key type of dictionary {keyType} for extracting a map". On snowflake side it is called a map, but on the C# side it is a dictionary. I'm using dictionary instead of map for purpose.